src/Workspaces/Core/Portable/Workspace/Solution/ProjectState.cs

```diff
@@ -547,9 +547,12 @@ public ProjectState WithParseOptions(ParseOptions options)
                 return this;
             }
 
+            var onlyPreprocessorDirectiveChange = ParseOptions != null &&
+                (_languageServices.SyntaxTreeFactory?.OptionsDifferOnlyByPreprocessorDirectives(options, ParseOptions) ?? false);
+
             return With(
                 projectInfo: ProjectInfo.WithParseOptions(options).WithVersion(Version.GetNewerVersion()),
-                documentStates: DocumentStates.UpdateStates(static (state, options) => state.UpdateParseOptions(options), options));
+                documentStates: DocumentStates.UpdateStates((state, options) => state.UpdateParseOptions(options, onlyPreprocessorDirectiveChange), options));
```

Comment:
```suggestion
                documentStates: DocumentStates.UpdateStates(static (state, optionsAndFlag) => state.UpdateParseOptions(optionsAndFlag.options, optionsAndFlag.onlyPreprocessorDirectiveChange), (options, onlyPreprocessorDirectiveChange)));
```

Reply:
Going to decline making this; it's far more verbose and saves a single lambda allocation per project which is vastly smaller than literally anything else happening in this path.

Reply:
sure.